File: source/clibCore/Models/Entity/addonModel.vb

Option Explicit On
Option Strict On

Imports System
Imports System.Collections.Generic
Imports System.Text
Imports Contensive.BaseClasses
Imports Contensive.Core.Controllers
Imports Newtonsoft.Json

Namespace Contensive.Core.Models.Entity
    Public Class addonModel
        Inherits baseModel
        '
        '====================================================================================================
        '-- const
        Public Const contentName As String = "add-ons" '<------ set content name
        Public Const contentTableName As String = "ccaggregatefunctions" '<------ set to tablename for the primary content (used for cache names)
        Private Shadows Const contentDataSource As String = "default" '<----- set to datasource if not default
        '
        '====================================================================================================
        ' -- instance properties
        '
        'Public Property ID As Integer
        'Public Property Active As Boolean
        Public Property Admin As Boolean
        Public Property ArgumentList As String
        Public Property AsAjax As Boolean
        'Public Property BlockDefaultStyles As Boolean
        Public Property BlockEditTools As Boolean
        'Public Property ccGuid As String
        Public Property CollectionID As Integer
        Public Property Content As Boolean
        Public Property ContentCategoryID As Integer
        'Public Property ContentControlID As Integer
        Public Property Copy As String
        Public Property CopyText As String
        'Public Property CreatedBy As Integer
        'Public Property CreateKey As Integer
        'Public Property CustomStylesFilename As String
        'Public Property DateAdded As Date
        Public Property DotNetClass As String
        Public Property EditArchive As Boolean
        Public Property EditBlank As Boolean
        Public Property EditSourceID As Integer
        Public Property Email As Boolean
        Public Property Filter As Boolean
        Public Property FormXML As String
        Public Property Help As String
        Public Property HelpLink As String
        Public Property IconFilename As String
        Public Property IconHeight As Integer
        Public Property IconSprites As Integer
        Public Property IconWidth As Integer
        Public Property InFrame As Boolean
        Public Property IsInline As Boolean
        Public Property JavaScriptBodyEnd As String
        Public Property JavaScriptOnLoad As String
        Public Property JSFilename As fieldTypeTextFile
        Public Property Link As String
        Public Property MetaDescription As String
        Public Property MetaKeywordList As String
        'Public Property ModifiedBy As Integer
        'Public Property ModifiedDate As Date
        'Public Property Name As String
        Public Property NavTypeID As Integer
        Public Property ObjectProgramID As String
        Public Property OnBodyEnd As Boolean
        Public Property OnBodyStart As Boolean
        Public Property OnNewVisitEvent As Boolean
        Public Property OnPageEndEvent As Boolean
        Public Property OnPageStartEvent As Boolean
        Public Property OtherHeadTags As String
        Public Property PageTitle As String
        Public Property ProcessInterval As Integer
        Public Property ProcessNextRun As Date
        Public Property ProcessRunOnce As Boolean
        Public Property ProcessServerKey As String
        Public Property RemoteAssetLink As String
        Public Property RemoteMethod As Boolean
        Public Property RobotsTxt As String
        Public Property ScriptingCode As String
        Public Property ScriptingEntryPoint As String
        Public Property ScriptingLanguageID As Integer
        Public Property ScriptingTimeout As String
        'Public Property SortOrder As String
        Public Property StylesFilename As fieldTypeTextFile
        Public Property Template As Boolean        '
        '====================================================================================================
        Public Overloads Shared Function add(cpCore As coreClass) As addonModel
            Return add(Of addonModel)(cpCore)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function add(cpCore As coreClass, ByRef callersCacheNameList As List(Of String)) As addonModel
            Return add(Of addonModel)(cpCore, callersCacheNameList)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function create(cpCore As coreClass, recordId As Integer) As addonModel
            Return create(Of addonModel)(cpCore, recordId)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function create(cpCore As coreClass, recordId As Integer, ByRef callersCacheNameList As List(Of String)) As addonModel
            Return create(Of addonModel)(cpCore, recordId, callersCacheNameList)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function create(cpCore As coreClass, recordGuid As String) As addonModel
            Return create(Of addonModel)(cpCore, recordGuid)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function create(cpCore As coreClass, recordGuid As String, ByRef callersCacheNameList As List(Of String)) As addonModel
            Return create(Of addonModel)(cpCore, recordGuid, callersCacheNameList)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function createByName(cpCore As coreClass, recordName As String) As addonModel
            Return createByName(Of addonModel)(cpCore, recordName)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function createByName(cpCore As coreClass, recordName As String, ByRef callersCacheNameList As List(Of String)) As addonModel
            Return createByName(Of addonModel)(cpCore, recordName, callersCacheNameList)
        End Function
        '
        '====================================================================================================
        Public Overloads Sub save(cpCore As coreClass)
            MyBase.save(cpCore)
        End Sub
        '
        '====================================================================================================
        Public Overloads Shared Sub delete(cpCore As coreClass, recordId As Integer)
            delete(Of addonModel)(cpCore, recordId)
        End Sub
        '
        '====================================================================================================
        Public Overloads Shared Sub delete(cpCore As coreClass, ccGuid As String)
            delete(Of addonModel)(cpCore, ccGuid)
        End Sub
        '
        '====================================================================================================
        Public Overloads Shared Function createList(cpCore As coreClass, sqlCriteria As String, sqlOrderBy As String, callersCacheNameList As List(Of String)) As List(Of addonModel)
            Return createList(Of addonModel)(cpCore, sqlCriteria, sqlOrderBy, callersCacheNameList)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function createList(cpCore As coreClass, sqlCriteria As String, sqlOrderBy As String) As List(Of addonModel)
            Return createList(Of addonModel)(cpCore, sqlCriteria, sqlOrderBy)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function createList(cpCore As coreClass, sqlCriteria As String) As List(Of addonModel)
            Return createList(Of addonModel)(cpCore, sqlCriteria)
        End Function
        '
        '====================================================================================================
        Public Overloads Sub invalidatePrimaryCache(cpCore As coreClass, recordId As Integer)
            invalidateCacheSingleRecord(Of addonModel)(cpCore, recordId)
        End Sub
        '
        '====================================================================================================
        Public Overloads Shared Function getRecordName(cpcore As coreClass, recordId As Integer) As String
            Return baseModel.getRecordName(Of addonModel)(cpcore, recordId)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function getRecordName(cpcore As coreClass, ccGuid As String) As String
            Return baseModel.getRecordName(Of addonModel)(cpcore, ccGuid)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function getRecordId(cpcore As coreClass, ccGuid As String) As Integer
            Return baseModel.getRecordId(Of addonModel)(cpcore, ccGuid)
        End Function
        '
        '====================================================================================================
        Public Overloads Shared Function createDefault(cpcore As coreClass) As addonModel
            Return createDefault(Of addonModel)(cpcore)
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' Get a list of addons to run on new visit
        ''' </summary>
        ''' <param name="cp"></param>
        ''' <param name="someCriteria"></param>
        ''' <returns></returns>
        Public Shared Function createList_OnNewVisitEvent(cpCore As coreClass, callersCacheNameList As List(Of String)) As List(Of addonModel)
            Dim result As New List(Of addonModel)
            Try
                result = createList(cpCore, "(OnNewVisitEvent<>0)")
            Catch ex As Exception
                cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' Get list of addons that run on page start event
        ''' </summary>
        ''' <param name="cp"></param>
        ''' <param name="someCriteria"></param>
        ''' <returns></returns>
        Public Shared Function createList_OnPageStartEvent(cpCore As coreClass, callersCacheNameList As List(Of String)) As List(Of addonModel)
            Dim result As New List(Of addonModel)
            Try
                result = createList(cpCore, "(OnPageStartEvent<>0)")
            Catch ex As Exception
                cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' pattern get a list of objects from this model
        ''' </summary>
        ''' <param name="cp"></param>
        ''' <param name="someCriteria"></param>
        ''' <returns></returns>
        Public Shared Function createList_RemoteMethods(cpCore As coreClass, callersCacheNameList As List(Of String)) As List(Of addonModel)
            Dim result As New List(Of addonModel)
            Try
                result = createList(cpCore, "(remoteMethod=1)")
            Catch ex As Exception
                cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        Public Shared Function createList_pageDependencies(cpCore As coreClass, pageId As Integer) As List(Of addonModel)
            Dim result As New List(Of addonModel)
            Try
                result = createList(cpCore, "(id in (select addonId from ccAddonPageRules where (pageId=" & pageId & ")))")
            Catch ex As Exception
                cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        Public Shared Function createList_templateDependencies(cpCore As coreClass, templateId As Integer) As List(Of addonModel)
            Dim result As New List(Of addonModel)
            Try
                result = createList(cpCore, "(id in (select addonId from ccAddonTemplateRules where (templateId=" & templateId & ")))")
            Catch ex As Exception
                cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        Public Class addonCacheClass
            Private dictIdAddon As New Dictionary(Of Integer, addonModel)
            Private dictGuidId As New Dictionary(Of String, Integer)
            Private dictNameId As New Dictionary(Of String, Integer)
            '
            Private onBodyEndIdList As New List(Of Integer)
            Private onBodyStartIdList As New List(Of Integer)
            Private onNewVisitIdList As New List(Of Integer)
            Private OnPageEndIdList As New List(Of Integer)
            Private OnPageStartIdList As New List(Of Integer)
            Public robotsTxt As String = ""
            '
            Public Sub add(addon As addonModel)
                If (Not dictIdAddon.ContainsKey(addon.id)) Then
                    If (Not dictGuidId.ContainsKey(addon.ccguid)) Then
                        If (Not dictNameId.ContainsKey(addon.name)) Then
                            dictIdAddon.Add(addon.id, addon)
                            dictGuidId.Add(addon.ccguid.ToLower(), addon.id)
                            dictNameId.Add(addon.name.ToLower(), addon.id)
                        End If
                    End If
                End If
                If (addon.OnBodyEnd) And (Not onBodyEndIdList.Contains(addon.id)) Then
                    onBodyEndIdList.Add(addon.id)
                End If
                If (addon.OnBodyStart) And (Not onBodyStartIdList.Contains(addon.id)) Then
                    onBodyStartIdList.Add(addon.id)
                End If
                If (addon.OnNewVisitEvent) And (Not onNewVisitIdList.Contains(addon.id)) Then
                    onNewVisitIdList.Add(addon.id)
                End If
                If (addon.OnPageEndEvent) And (Not OnPageEndIdList.Contains(addon.id)) Then
                    OnPageEndIdList.Add(addon.id)
                End If
                If (addon.OnPageStartEvent) And (Not OnPageStartIdList.Contains(addon.id)) Then
                    OnPageStartIdList.Add(addon.id)
                End If
                robotsTxt &= vbCrLf & addon.RobotsTxt
            End Sub
            Public Function getAddonByGuid(guid As String) As addonModel
                If (Me.dictGuidId.ContainsKey(guid.ToLower())) Then
                    Return getAddonById(Me.dictGuidId.Item(guid.ToLower()))
                End If
                Return Nothing
            End Function
            Public Function getAddonByName(name As String) As addonModel
                If (Me.dictNameId.ContainsKey(name.ToLower())) Then
                    Return getAddonById(Me.dictNameId.Item(name.ToLower()))
                End If
                Return Nothing
            End Function
            Public Function getAddonById(addonId As Integer) As addonModel
                If (Me.dictIdAddon.ContainsKey(addonId)) Then
                    Return Me.dictIdAddon.Item(addonId)
                End If
                Return Nothing
            End Function
            '
            Private Function getAddonList(addonIdList As List(Of Integer)) As List(Of addonModel)
                Dim result As New List(Of addonModel)
                For Each addonId As Integer In addonIdList
                    result.Add(getAddonById(addonId))
                Next
                Return result
            End Function
            '
            Public Function getOnBodyEndAddonList() As List(Of addonModel)
                Return getAddonList(onBodyEndIdList)
            End Function
            '
            Public Function getOnBodyStartAddonList() As List(Of addonModel)
                Return getAddonList(onBodyStartIdList)
            End Function
            '
            Public Function getOnNewVisitAddonList() As List(Of addonModel)
                Return getAddonList(onNewVisitIdList)
            End Function
            '
            Public Function getOnPageEndAddonList() As List(Of addonModel)
                Return getAddonList(OnPageEndIdList)
            End Function
            '
            Public Function getOnPageStartAddonList() As List(Of addonModel)
                Return getAddonList(OnPageStartIdList)
            End Function
        End Class
    End Class
End Namespace

File: source/clibCoreCs/Models/Entity/addonModel.cs

using System;
using System.Reflection;
using System.Xml;
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using Contensive.Core;
using Contensive.Core.Models.Entity;
using Contensive.Core.Controllers;
using static Contensive.Core.Controllers.genericController;
using static Contensive.Core.constants;
//
namespace Contensive.Core.Models.Entity {
    public class addonModel : baseModel {
        //
        //====================================================================================================
        //-- const
        public const string contentName = "add-ons";
        public const string contentTableName = "ccaggregatefunctions";
        private const string contentDataSource = "default";
        //
        //====================================================================================================
        // -- instance properties
        //
        public bool Admin { get; set; }
        public string ArgumentList { get; set; }
        public bool AsAjax { get; set; }
        public bool BlockEditTools { get; set; }
        public int CollectionID { get; set; }
        public bool Content { get; set; }
        public string Copy { get; set; }
        public string CopyText { get; set; }
        public string DotNetClass { get; set; }
        public bool Email { get; set; }
        public bool Filter { get; set; }
        public string FormXML { get; set; }
        public string Help { get; set; }
        public string HelpLink { get; set; }
        public string IconFilename { get; set; }
        public int IconHeight { get; set; }
        public int IconSprites { get; set; }
        public int IconWidth { get; set; }
        public bool InFrame { get; set; }
        public bool IsInline { get; set; }
        public string JavaScriptBodyEnd { get; set; }
        //Public Property JavaScriptOnLoad As String
        public string JSHeadScriptSrc { get; set; }
        public string JSBodyScriptSrc { get; set; }
        public fieldTypeJavascriptFile JSFilename { get; set; }
        public string Link { get; set; }
        public string MetaDescription { get; set; }
        public string MetaKeywordList { get; set; }
        public int NavTypeID { get; set; }
        public string ObjectProgramID { get; set; }
        public bool OnBodyEnd { get; set; }
        public bool OnBodyStart { get; set; }
        public bool OnNewVisitEvent { get; set; }
        public bool OnPageEndEvent { get; set; }
        public bool OnPageStartEvent { get; set; }
        public bool htmlDocument { get; set; }
        public string OtherHeadTags { get; set; }
        public string PageTitle { get; set; }
        public int ProcessInterval { get; set; }
        public DateTime ProcessNextRun { get; set; }
        public bool ProcessRunOnce { get; set; }
        public string ProcessServerKey { get; set; }
        public string RemoteAssetLink { get; set; }
        public bool RemoteMethod { get; set; }
        public string RobotsTxt { get; set; }
        public string ScriptingCode { get; set; }
        public string ScriptingEntryPoint { get; set; }
        public int ScriptingLanguageID { get; set; }
        public string ScriptingTimeout { get; set; }
        public fieldTypeCSSFile StylesFilename { get; set; }
        public string StylesLinkHref { get; set; }
        public bool Template { get; set; }
        //====================================================================================================
        public static addonModel add(coreClass cpCore) {
            return add<addonModel>(cpCore);
        }
        //
        //====================================================================================================
        public static addonModel add(coreClass cpCore, ref List<string> callersCacheNameList) {
            return add<addonModel>(cpCore, ref callersCacheNameList);
        }
        //
        //====================================================================================================
        public static addonModel create(coreClass cpCore, int recordId) {
            addonModel result = create<addonModel>(cpCore, recordId);
            if (result != null) {
                if (string.IsNullOrEmpty(result.ccguid)) {
                    result.ccguid = genericController.createGuid();
                }
            }
            return result;
        }
        //
        //====================================================================================================
        public static addonModel create(coreClass cpCore, int recordId, ref List<string> callersCacheNameList) {
            addonModel result = create<addonModel>(cpCore, recordId, ref callersCacheNameList);
            if (result != null) {
                if (string.IsNullOrEmpty(result.ccguid)) {
                    result.ccguid = genericController.createGuid();
                }
            }
            return result;
        }
        //
        //====================================================================================================
        public static addonModel create(coreClass cpCore, string recordGuid) {
            return create<addonModel>(cpCore, recordGuid);
        }
        //
        //====================================================================================================
        public static addonModel create(coreClass cpCore, string recordGuid, ref List<string> callersCacheNameList) {
            return create<addonModel>(cpCore, recordGuid, ref callersCacheNameList);
        }
        //
        //====================================================================================================
        public static addonModel createByName(coreClass cpCore, string recordName) {
            return createByName<addonModel>(cpCore, recordName);
        }
        //
        //====================================================================================================
        public static addonModel createByName(coreClass cpCore, string recordName, ref List<string> callersCacheNameList) {
            return createByName<addonModel>(cpCore, recordName, ref callersCacheNameList);
        }
        //
        //====================================================================================================
        public void save(coreClass cpCore) {
            base.save(cpCore);
        }
        //
        //====================================================================================================
        public static void delete(coreClass cpCore, int recordId) {
            delete<addonModel>(cpCore, recordId);
        }
        //
        //====================================================================================================
        public static void delete(coreClass cpCore, string ccGuid) {
            delete<addonModel>(cpCore, ccGuid);
        }
        //
        //====================================================================================================
        public static List<addonModel> createList(coreClass cpCore, string sqlCriteria, string sqlOrderBy, List<string> callersCacheNameList) {
            return createList<addonModel>(cpCore, sqlCriteria, sqlOrderBy, callersCacheNameList);
        }
        //
        //====================================================================================================
        public static List<addonModel> createList(coreClass cpCore, string sqlCriteria, string sqlOrderBy) {
            return createList<addonModel>(cpCore, sqlCriteria, sqlOrderBy);
        }
        //
        //====================================================================================================
        public static List<addonModel> createList(coreClass cpCore, string sqlCriteria) {
            return createList<addonModel>(cpCore, sqlCriteria);
        }
        //
        //====================================================================================================
        public static void invalidateCache(coreClass cpCore, int recordId) {
            invalidateCacheSingleRecord<addonModel>(cpCore, recordId);
            Models.Complex.routeDictionaryModel.invalidateCache(cpCore);
        }
        //
        //====================================================================================================
        public static string getRecordName(coreClass cpcore, int recordId) {
            return baseModel.getRecordName<addonModel>(cpcore, recordId);
        }
        //
        //====================================================================================================
        public static string getRecordName(coreClass cpcore, string ccGuid) {
            return baseModel.getRecordName<addonModel>(cpcore, ccGuid);
        }
        //
        //====================================================================================================
        public static int getRecordId(coreClass cpcore, string ccGuid) {
            return baseModel.getRecordId<addonModel>(cpcore, ccGuid);
        }
        //
        //====================================================================================================
        public static addonModel createDefault(coreClass cpcore) {
            return createDefault<addonModel>(cpcore);
        }
        //
        //====================================================================================================
        /// <summary>
        /// Get a list of addons to run on new visit
        /// </summary>
        /// <param name="cp"></param>
        /// <param name="someCriteria"></param>
        /// <returns></returns>
        public static List<addonModel> createList_OnNewVisitEvent(coreClass cpCore, List<string> callersCacheNameList) {
            List<addonModel> result = new List<addonModel>();
            try {
                result = createList(cpCore, "(OnNewVisitEvent<>0)");
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        /// <summary>
        /// Get list of addons that run on page start event
        /// </summary>
        /// <param name="cp"></param>
        /// <param name="someCriteria"></param>
        /// <returns></returns>
        public static List<addonModel> createList_OnPageStartEvent(coreClass cpCore, List<string> callersCacheNameList) {
            List<addonModel> result = new List<addonModel>();
            try {
                result = createList(cpCore, "(OnPageStartEvent<>0)");
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        /// <summary>
        /// pattern get a list of objects from this model
        /// </summary>
        /// <param name="cp"></param>
        /// <param name="someCriteria"></param>
        /// <returns></returns>
        public static List<addonModel> createList_RemoteMethods(coreClass cpCore, List<string> callersCacheNameList) {
            List<addonModel> result = new List<addonModel>();
            try {
                result = createList(cpCore, "(remoteMethod=1)");
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return result;
        }
        //
        public static List<addonModel> createList_pageDependencies(coreClass cpCore, int pageId) {
            List<addonModel> result = new List<addonModel>();
            try {
                result = createList(cpCore, "(id in (select addonId from ccAddonPageRules where (pageId=" + pageId + ")))");
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return result;
        }
        //
        public static List<addonModel> createList_templateDependencies(coreClass cpCore, int templateId) {
            List<addonModel> result = new List<addonModel>();
            try {
                result = createList(cpCore, "(id in (select addonId from ccAddonTemplateRules where (templateId=" + templateId + ")))");
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return result;
        }
        //
        public class addonCacheClass {
            private Dictionary<int, addonModel> dictIdAddon = new Dictionary<int, addonModel>();
            private Dictionary<string, int> dictGuidId = new Dictionary<string, int>();
            private Dictionary<string, int> dictNameId = new Dictionary<string, int>();
            //
            private List<int> onBodyEndIdList = new List<int>();
            private List<int> onBodyStartIdList = new List<int>();
            private List<int> onNewVisitIdList = new List<int>();
            private List<int> OnPageEndIdList = new List<int>();
            private List<int> OnPageStartIdList = new List<int>();
            public string robotsTxt = "";
            //
            public void add(coreClass cpcore, addonModel addon) {
                if (!dictIdAddon.ContainsKey(addon.id)) {
                    dictIdAddon.Add(addon.id, addon);
                    if (string.IsNullOrEmpty(addon.ccguid)) {
                        addon.ccguid = genericController.createGuid();
                        addon.save(cpcore);
                    }
                    if (!dictGuidId.ContainsKey(addon.ccguid.ToLower())) {
                        dictGuidId.Add(addon.ccguid.ToLower(), addon.id);
                        if (string.IsNullOrEmpty(addon.name.Trim())) {
                            addon.name = "addon " + addon.id.ToString();
                            addon.save(cpcore);
                        }
                        if (!dictNameId.ContainsKey(addon.name.ToLower())) {
                            dictNameId.Add(addon.name.ToLower(), addon.id);
                        }
                    }
                }
                if ((addon.OnBodyEnd) && (!onBodyEndIdList.Contains(addon.id))) {
                    onBodyEndIdList.Add(addon.id);
                }
                if ((addon.OnBodyStart) && (!onBodyStartIdList.Contains(addon.id))) {
                    onBodyStartIdList.Add(addon.id);
                }
                if ((addon.OnNewVisitEvent) && (!onNewVisitIdList.Contains(addon.id))) {
                    onNewVisitIdList.Add(addon.id);
                }
                if ((addon.OnPageEndEvent) && (!OnPageEndIdList.Contains(addon.id))) {
                    OnPageEndIdList.Add(addon.id);
                }
                if ((addon.OnPageStartEvent) && (!OnPageStartIdList.Contains(addon.id))) {
                    OnPageStartIdList.Add(addon.id);
                }
                robotsTxt += "\r\n" + addon.RobotsTxt;
            }
            public addonModel getAddonByGuid(string guid) {
                if (this.dictGuidId.ContainsKey(guid.ToLower())) {
                    return getAddonById(this.dictGuidId[guid.ToLower()]);
                }
                return null;
            }
            public addonModel getAddonByName(string name) {
                if (this.dictNameId.ContainsKey(name.ToLower())) {
                    return getAddonById(this.dictNameId[name.ToLower()]);
                }
                return null;
            }
            public addonModel getAddonById(int addonId) {
                if (this.dictIdAddon.ContainsKey(addonId)) {
                    return this.dictIdAddon[addonId];
                }
                return null;
            }
            //
            private List<addonModel> getAddonList(List<int> addonIdList) {
                List<addonModel> result = new List<addonModel>();
                foreach (int addonId in addonIdList) {
                    result.Add(getAddonById(addonId));
                }
                return result;
            }
            //
            public List<addonModel> getOnBodyEndAddonList() {
                return getAddonList(onBodyEndIdList);
            }
            //
            public List<addonModel> getOnBodyStartAddonList() {
                return getAddonList(onBodyStartIdList);
            }
            //
            public List<addonModel> getOnNewVisitAddonList() {
                return getAddonList(onNewVisitIdList);
            }
            //
            public List<addonModel> getOnPageEndAddonList() {
                return getAddonList(OnPageEndIdList);
            }
            //
            public List<addonModel> getOnPageStartAddonList() {
                return getAddonList(OnPageStartIdList);
            }
        }
    }
}
